core/src/main/java/com/datastrato/gravitino/storage/relational/service/CatalogMetaService.java

```diff
@@ -0,0 +1,206 @@
+            if (overwrite) {
+              mapper.insertCatalogMetaOnDuplicateKeyUpdate(po);
+            } else {
+              mapper.insertCatalogMeta(po);
+            }
+          });
+    } catch (RuntimeException re) {
+      if (re.getCause() != null
+          && re.getCause().getCause() != null
+          && re.getCause().getCause() instanceof SQLIntegrityConstraintViolationException) {
+        // TODO We should make more fine-grained exception judgments
+        // Usually throwing `SQLIntegrityConstraintViolationException` means that
+        // SQL violates the constraints of `primary key` and `unique key`.
+        // We simply think that the entity already exists at this time.
+        throw new EntityAlreadyExistsException(
+            String.format("Catalog entity: %s already exists", catalogEntity.nameIdentifier()));
+      }
+      throw re;
+    }
+  }
+
+  public <E extends Entity & HasIdentifier> CatalogEntity updateCatalog(
+      NameIdentifier identifier, Function<E, E> updater) throws IOException {
+    NameIdentifier.checkCatalog(identifier);
+    String metalakeName = identifier.namespace().level(0);
+    String catalogName = identifier.name();
+    Long metalakeId =
+        SessionUtils.getWithoutCommit(
+            MetalakeMetaMapper.class, mapper -> mapper.selectMetalakeIdMetaByName(metalakeName));
+    if (metalakeId == null) {
+      throw new NoSuchEntityException(
+          NoSuchEntityException.NO_SUCH_ENTITY_MESSAGE, identifier.namespace().toString());
+    }
+
+    CatalogPO oldCatalogPO =
+        SessionUtils.getWithoutCommit(
+            CatalogMetaMapper.class,
+            mapper -> mapper.selectCatalogMetaByMetalakeIdAndName(metalakeId, catalogName));
+    if (oldCatalogPO == null) {
+      throw new NoSuchEntityException(
+          NoSuchEntityException.NO_SUCH_ENTITY_MESSAGE, identifier.toString());
+    }
+
+    CatalogEntity oldCatalogEntity =
+        POConverters.fromCatalogPO(oldCatalogPO, identifier.namespace());
+    CatalogEntity newEntity = (CatalogEntity) updater.apply((E) oldCatalogEntity);
+    Preconditions.checkArgument(
+        Objects.equals(oldCatalogEntity.id(), newEntity.id()),
+        "The updated catalog entity id: %s should be same with the catalog entity id before: %s",
+        newEntity.id(),
+        oldCatalogEntity.id());
+    Integer updateResult;
+    try {
+      updateResult =
+          SessionUtils.doWithCommitAndFetchResult(
+              CatalogMetaMapper.class,
+              mapper ->
+                  mapper.updateCatalogMeta(
+                      POConverters.updateCatalogPOWithVersion(oldCatalogPO, newEntity, metalakeId),
+                      oldCatalogPO));
+    } catch (RuntimeException re) {
+      if (re.getCause() != null
+          && re.getCause().getCause() != null
+          && re.getCause().getCause() instanceof SQLIntegrityConstraintViolationException) {
+        // TODO We should make more fine-grained exception judgments
```

Comment:
It would be best to create an issue to track it at this point. The current implementation is not aesthetically pleasing.

Reply:
Had created the issue before: #2227